xla/service/collective_pipeliner_test.cc

```diff
@@ -1348,6 +1348,86 @@ ENTRY entry {
   EXPECT_EQ(add_instr_loop->opcode(), HloOpcode::kAdd);
 }
 
+TEST_F(CollectivePipelinerTest,
+       TransformIncrementIndexByOneStartFromOneBackwards) {
+  constexpr absl::string_view hlo_string = R"(
+HloModule module
+
+add {
+  lhs = bf16[] parameter(0)
+  rhs = bf16[] parameter(1)
+  ROOT add = bf16[] add(lhs, rhs)
+}
+
+while_cond {
+  param = (s32[], bf16[3,8,128], bf16[3,1,2,128]) parameter(0)
+  loop_index = s32[] get-tuple-element(param), index=0
+  c4 = s32[] constant(4)
+  ROOT cmp = pred[] compare(loop_index, c4), direction=LT
+}
+
+while_body {
+  param = (s32[], bf16[3,8,128], bf16[3,1,2,128]) parameter(0)
+  loop_index = s32[] get-tuple-element(param), index=0
+  partial_output = bf16[3,8,128] get-tuple-element(param), index=1
+  slice_input = bf16[3,1,2,128] get-tuple-element(param), index=2
+  c0 = s32[] constant(0)
+  c1 = s32[] constant(1)
+  next_loop_index = s32[] add(loop_index, c1)
+  c4 = s32[] constant(4)
+  four_minus_loop_index = s32[] subtract(c4, loop_index)
+  dynamic-slice = bf16[1,1,2,128] dynamic-slice(slice_input, four_minus_loop_index, c0, c0, c0), dynamic_slice_sizes={1,1,2,128}
+  dynamic-slice_reshape = bf16[1,2,128] reshape(dynamic-slice)
+  add = bf16[1,2,128] add(dynamic-slice_reshape, dynamic-slice_reshape), control-predecessors={c4}
+  all-gather = bf16[1,8,128] all-gather(add), dimensions={1}, replica_groups={}
+  updated_partial_output = bf16[3,8,128] dynamic-update-slice(partial_output, all-gather, four_minus_loop_index, c0, c0)
+  ROOT tuple = (s32[], bf16[3,8,128], bf16[3,1,2,128]) tuple(next_loop_index, updated_partial_output, slice_input), control-predecessors={add}
+}
+
+ENTRY entry {
+  c1 = s32[] constant(1)
+  p0 = bf16[3,8,128] parameter(0)
+  p1 = bf16[3,1,2,128] parameter(1)
+  tuple = (s32[], bf16[3,8,128], bf16[3,1,2,128]) tuple(c1, p0, p1)
+  while = (s32[], bf16[3,8,128], bf16[3,1,2,128]) while(tuple), condition=while_cond, body=while_body
+  ROOT gte1 = bf16[3,8,128] get-tuple-element(while), index=1
+}
+)";
+  auto module = ParseAndReturnUnverifiedModule(hlo_string, config_).value();
+  EXPECT_TRUE(RunOptimizer(module.get(), /*last_run=*/true, 0,
+                           /*pipeline_use_tree=*/false,
+                           /*process_different_sized_ops=*/false,
+                           CollectivePipeliner::PipeliningDirection::kBackward,
+                           IsAllGather)
+                  .value());
+  XLA_VLOG_LINES(1, module->ToString());
+  const int64_t while_count = absl::c_count_if(
+      module->entry_computation()->instructions(),
+      [](const HloInstruction* instruction) {
+        return HloPredicateIsOp<HloOpcode::kWhile>(instruction);
+      });
+  EXPECT_EQ(while_count, 1);
```

Comment:
Can you rewrite this part to filecheck? All you want to check is the increment so it should not be long, I think.

Reply:
I removed some of the verifications duplicated in other tests but didn't switch to FileCheck since e.g. TransformIncrementIndexByOneBackwards doesn't use it either. Let me know if you want to use it anyway.

Reply:
Without so many assertions I think in practice it's now less mandated :)